doc/config.txt
*****************
Configuring Hooke
*****************

Hooke initializes a number of variables by reading configuration
files.

Syntax
------

Hooke uses Python's configparser_ library to read the config files.
The format is similar to MS Windows' INI (initialization) format::

    [My Section]
    foo = bar
    baz = 5

.. _configparser: http://docs.python.org/library/configparser.html

Examples
--------

Run::

    $ ./bin/hk.py --save-config -c ''

or::

    $ ./bin/hk.py -c 'save_config --output .hooke.cfg'

To write a well commented example config file to :file:`.hooke.cfg`.


Finding configuration files
---------------------------

The default search path follows the `Filesystem Hierarchy Standard`_,
and so will probably need adjustment for non-\*nix systems.  The
default path list is

* :file:`/usr/share/hooke/hooke.cfg`
* :file:`/etc/hooke/hooke.cfg`
* :file:`~/.hooke.cfg`
* :file:`./.hooke.cfg`

but alternatives can be specified from the command line launching
Hooke.

.. _Filesystem Hierarchy Standard: http://www.pathname.com/fhs/

.. todo:: Implement command-line specified config file.

.. warning:: Any changes you make to the configuration during a Hooke
  run will *not be saved* when Hooke exits.  To save the changes,
  either run the `save config` command before closing Hooke, or start
  Hooke with the ``--save-config`` option.


pca_config.txt
1,3,6,7,8,9x15,10,11
0.000008
1,2,3,4,5,6,7,8,9,10,11

max 12 colonne !!!



str(peak_number)+     # non considerato
str(delta_mean)+      # 0
str(delta_median)+    # 1 -
str(force_mean)+      # 2
str(force_median)+    # 3 -
str(first_peak_cl)+   # 4 -
str(last_peak_cl)+    # 5 -
str(max_force)+       # 6
str(min_force)+       # 7
str(max_delta)+       # 8
str(min_delta)+       # 9
str(delta_stdev)+     # 10
str(forces_stdev)+    # 11

multipca dataset_gb1x8_coor_blind50.txt 9

CHANGELOG
THIS CHANGELOG FILE CONTANS PRE-SVN CHANGELOGS AND SUMMARIES OF POST-SVN CHANGELOGS.
FROM 0.8.4 ONWARD, DETAILED CHANGELOGS ARE AVAILABLE ON THE HOOKE SUBVERSION REPOSITORY.

0.8.4
(2008-x-x)
    PLUGINS:
    macro.py:
        hooke does not crash if it doesn't have permissions to create the folder
	fixed overwriting of export for curves with the same name and different numerical extension
	fixed macrodir (out of try-catch)
	cleaned debug output in execmacro
    DRIVERS:
	alternative version of picoforce driver
    INPUT
	merged (partially) libinput (so far in hooke_cli)
    OUTLET	
	merged liboutlet


0.8.3
(2008-04-16)
    PLUGINS:
    generalvclamp.py:
        fixed autopeak header
        fixed autopeak slope (now unwanted slope values are discarded)

0.8.2
(2008-04-10)
    PLUGINS:
    flatfilts.py:
        convfilt does not crash if a file is not a curve
    generalvclamp.py:
        autopeak now saves curve data correctly
        autopeak now generates a dummy note (so that copylog/notelog is aware you measured the curve)

0.8.1
(2008-04-07)
    PLUGINS:
    generalvclamp.py:
        fixed DeprecationWarning in flatten
    flatfilts.py
        convfilt now working


0.8.0:
(2008-04-04)
    hooke.py:
        sanity check of CLI plugins to avoid function overloading at startup
    hooke_cli.py ; libhooke.py:
        now playlists keep the index (when you reload the playlist, it starts from the
            last observed curve)
        updated plot to use _send_plot()
        hooke.conf accepts lists as arguments for variables in <display>
        txt, export now have consistent argument order (thanks to A.G.Casado for pointing me that)
        txt crashes no more if no filename is given (thanks to A.G.Casado for pointing me that)
    libhookecurve.py:
        added add_set() , remove_set() methods to make life easier for plugin writers
    procplots.py:
        plotmanip_correct() works with new picoforce.py deflection output (see)
    PLUGINS:
    fit.py:
        updated wlc to use _send_plot()
        wlc noauto now keeps the contact point
        wlc reclick to click again the contact point
        temperature now set in hooke.conf
    generalvclamp.py:
        implemented slope (thanks to Marco Brucale)
        implemented autopeak
    flatfilts.py:
        convfilt,peaks use flattened curve
    macro.py:
        (new) added macro plugin (thanks to Alberto Gomez Casado)
    DRIVERS:
    picoforce.py:
        fixed trigger bug! (thanks to Alberto Gomez Casado)
        better deflection output (separated extension,retraction)

0.7.5:
(2008-03-27)
    hooke_cli.py:
        removed outdated size command
    PLUGINS:
    generalvclamp.py:
        implemented flatten
    DRIVERS:
    added tutorialdriver.py driver
    csvdriver.py:
        fixed (forgot close_all() method)

0.7.4:
(2008-03-19)
    added csvdriver driver
    hooke_cli.py:
        fixed plot manipulators handling (now it's safe to comment a
              plot manipulator on hooke.conf)
    PLUGINS:
    fit.py:
        fixed possible crash when clicking two times the same point on wlc

0.7.3:
(2008-01-10)
    hooke_cli.py:
        fixed crash on copylog
    PLUGINS:
        massanalysis.py:
            Initial release
        tutorial.py:
            Tutorial plugin, initial release

0.7.2.1:
(2007-11-30)
    PLUGINS:
        flatfilt.py:
            fixed crash on Windows

0.7.2:
(2007-11-29)
    hooke.py:
        new configuration variable hookedir
    hooke_cli.py:
        copylog now checks if the destination is a real directory
        fixed crashes in set
    PLUGINS:
        generalvclamp.py:
            fixed a crash in forcebase when picking two times the same point
        flatfilt.py:
            fixed crash due to convfilt.conf impossible to load
            initial implementation of the blind window for convfilt
            initial data set maps (NOT FINISHED)

0.7.1:
(2007-11-26)
    PLUGINS:
        flatfilts.py:
            fixed possible crash in convfilt
            implemented configuration file convfilt.conf
            convfilt defaults are now 5 peaks 5 times more the noise absdev
            implemented convconf
            implemented setconf
        libpeakspot.py:
            fixed:now it really uses noise_absdev

0.7.0:
(2007-11-15)
    hooke_cli.py:
        implemented _send_plot() helper API function
    PLUGINS:
        generalvclamp.py:
            fixed forcebase to work with subtplot
        flatfilts.py:
            implemented convfilt!
            added libpeakspot.py (helping library for convolution filter)

0.6.5:
(2007-11-06)
    hooke_cli.py, hooke.py:
        plateau and contact (unmaintained) deleted and scheduled for re-release in generalvramp
        implemented _measure_N_points()
    PLUGINS:
    generalvclamp.py:
        implemented forcebase
    fit.py:
        wlc now accepts and uses temperature as an argument
        wlc has been cleaned and uses new APIs

0.6.4:
(2007-10-23)
    hooke_cli.py, libhooke.py:
        implemented support for defining order of plotmanip methods in hooke.conf
    hooke_cli.py:
        implemented delta
        implemented point
        attempted fix to bug 0033 (notelog crashing Hooke when using Unicode characters)
    PLUGINS:
    generalvramp.py:
        began to move velocity ramp force spectroscopy-specific things in separate plugin
    procplots.py:
        added detriggerize; "set detrigger" 0/1 disables/enables it.
    DRIVERS:
    picoforce.py:
        removed detriggerize() from driver

0.6.3:
(2007-10-02)
    hooke_cli.py:
        rewritten txt command, now working
    DRIVERS:
    picoforce.py:
        implemented detriggerize() to bypass the Picoforce trigger bug
    PLUGINS:
    superimpose.py:
        implemented plotavgimpose

0.6.2:
(2007-09-27)
    hooke_cli.py:
        fixed error handling in notelog
        smarter handling of directory names in genlist
        unexpected error handling in do_plot()
    hooke.py:

hooke/config.py
# Copyright (C) 2010 W. Trevor King <wking@drexel.edu>
#
# This file is part of Hooke.
#
# Hooke is free software: you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# Hooke is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General
# Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with Hooke.  If not, see
# <http://www.gnu.org/licenses/>.

"""Configuration defaults, read/write, and template file creation for
Hooke.
"""

import ConfigParser as configparser
import logging
import os.path
import textwrap
import unittest

from .compat.odict import odict as OrderedDict
from .util.convert import to_string, from_string


DEFAULT_PATHS = [
    '/usr/share/hooke/hooke.cfg',
    '/etc/hooke/hooke.cfg',
    '~/.hooke.cfg',
    '.hooke.cfg',
    ]
"""We start with the system files, and work our way to the more
specific user files, so the user can override the sysadmin who
in turn overrides the developer defaults.
"""

class Setting (object):
    """An entry (section or option) in HookeConfigParser.
    """
    def __init__(self, section, option=None, value=None, type='string',
                 count=1, help=None, wrap=True):
        self.section = section
        self.option = option
        self.value = value
        self.type = type
        self.count = count
        self.help = help
        self.wrap = wrap

    def __eq__(self, other):
        for attr in ['__class__', 'section', 'option', 'value', 'help']:
            value = getattr(self, attr)
            o_value = getattr(other, attr)
            if o_value != value:
                return False
        return True

    def is_section(self):
        return self.option == None

    def is_option(self):
        return not self.is_section()

    def write(self, fp, value=None, comments=True, wrapper=None):
        if comments == True and self.help != None:
            text = self.help
            if self.wrap == True:
                if wrapper == None:
                    wrapper = textwrap.TextWrapper(
                        initial_indent='# ',
                        subsequent_indent='# ')
                text = wrapper.fill(text)
            else:
                text = '# ' + '\n# '.join(text.splitlines())
            fp.write(text.rstrip()+'\n')
        if self.is_section():
            fp.write("[%s]\n" % self.section)
        else:
            fp.write("%s = %s\n" % (self.option,
                                    str(value).replace('\n', '\n\t')))

DEFAULT_SETTINGS = [
    Setting('conditions', help='Default environmental conditions in case they are not specified in the force curve data.  Configuration options in this section are available to every plugin.'),
    Setting('conditions', 'temperature', value='301', type='float', help='Temperature in Kelvin'),
    # Logging settings
    Setting('loggers', help='Configure loggers, see\nhttp://docs.python.org/library/logging.html#configuration-file-format', wrap=False),
    Setting('loggers', 'keys', 'root, hooke', help='Hooke only uses the hooke logger, but other included modules may also use logging and you can configure their loggers here as well.'),
    Setting('handlers', help='Configure log handlers, see\nhttp://docs.python.org/library/logging.html#configuration-file-format', wrap=False),
    Setting('handlers', 'keys', 'hand1'),
    Setting('formatters', help='Configure log formatters, see\nhttp://docs.python.org/library/logging.html#configuration-file-format', wrap=False),
    Setting('formatters', 'keys', 'form1'),
    Setting('logger_root', help='Configure the root logger, see\nhttp://docs.python.org/library/logging.html#configuration-file-format', wrap=False),
    Setting('logger_root', 'level', 'NOTSET'),
    Setting('logger_root', 'handlers', 'hand1'),
    Setting('logger_hooke', help='Configure the hooke logger, see\nhttp://docs.python.org/library/logging.html#configuration-file-format', wrap=False),
    Setting('logger_hooke', 'level', 'DEBUG'),
    Setting('logger_hooke', 'handlers', 'hand1', help='No specific handlers here, just propagate up to the root logger'),
    Setting('logger_hooke', 'propagate', '0'),
    Setting('logger_hooke', 'qualname', 'hooke'),
    Setting('handler_hand1', help='Configure the default log handler, see\nhttp://docs.python.org/library/logging.html#configuration-file-format', wrap=False),
    Setting('handler_hand1', 'class', 'StreamHandler'),
    Setting('handler_hand1', 'level', 'WARN'),
    Setting('handler_hand1', 'formatter', 'form1'),
    Setting('handler_hand1', 'args', '(sys.stderr,)'),
    Setting('formatter_form1', help='Configure the default log formatter, see\nhttp://docs.python.org/library/logging.html#configuration-file-format', wrap=False),
    Setting('formatter_form1', 'format', '%(asctime)s %(levelname)s %(message)s'),
    Setting('formatter_form1', 'datefmt', '', help='Leave blank for ISO8601, e.g. "2003-01-23 00:29:50,411".'),
    Setting('formatter_form1', 'class', 'logging.Formatter'),
    ]

def get_setting(settings, match):
    """Return the first Setting object matching both match.section and
    match.option.
    """
    for s in settings:
        if s.section == match.section and s.option == match.option:
            return s
    return None

class HookeConfigParser (configparser.RawConfigParser):
    """A wrapper around configparser.RawConfigParser.

    You will probably only need .read and .write.

    Examples
    --------

    >>> import pprint
    >>> import sys
    >>> c = HookeConfigParser(default_settings=DEFAULT_SETTINGS)
    >>> c.write(sys.stdout) # doctest: +ELLIPSIS
    # Default environmental conditions in case they are not specified in
    # the force curve data.  Configuration options in this section are
    # available to every plugin.
    [conditions]
    # Temperature in Kelvin
    temperature = 301
    <BLANKLINE>
    # Configure loggers, see
    # http://docs.python.org/library/logging.html#configuration-file-format
    [loggers]
    # Hooke only uses the hooke logger, but other included modules may
    # also use logging and you can configure their loggers here as well.
    keys = root, hooke
    ...

    class:`HookeConfigParser` automatically converts typed settings.

    >>> section = 'test conversion'
    >>> c = HookeConfigParser(default_settings=[
    ...         Setting(section),
    ...         Setting(section, option='my string', value='Lorem ipsum', type='string'),
    ...         Setting(section, option='my bool', value=True, type='bool'),
    ...         Setting(section, option='my int', value=13, type='int'),
    ...         Setting(section, option='my float', value=3.14159, type='float'),
    ...         ])
    >>> pprint.pprint(c.items(section))  # doctest: +ELLIPSIS
    [('my string', u'Lorem ipsum'),
     ('my bool', True),
     ('my int', 13),
     ('my float', 3.1415...)]

    However, the regular `.get()` is not typed.  Users are encouraged
    to use the standard `.get*()` methods.

    >>> c.get('test conversion', 'my bool')
    u'True'
    >>> c.getboolean('test conversion', 'my bool')
    True
    """
    def __init__(self, paths=None, default_settings=None, defaults=None,
                 dict_type=OrderedDict, indent='# ', **kwargs):
        # Can't use super() because RawConfigParser is a classic class
        #super(HookeConfigParser, self).__init__(defaults, dict_type)
        configparser.RawConfigParser.__init__(self, defaults, dict_type)
        if paths == None:
            paths = []
        self._config_paths = paths
        if default_settings == None:
            default_settings = []
        self._default_settings = default_settings
        self._default_settings_dict = {}
        for key in ['initial_indent', 'subsequent_indent']:
            if key not in kwargs:
                kwargs[key] = indent
        self._comment_textwrap = textwrap.TextWrapper(**kwargs)
        self._setup_default_settings()

    def _setup_default_settings(self):
        for setting in self._default_settings: #reversed(self._default_settings):
            # reversed cause: http://docs.python.org/library/configparser.html
            # "When adding sections or items, add them in the reverse order of
            # how you want them to be displayed in the actual file."
            self._default_settings_dict[
                (setting.section, setting.option)] = setting
            if setting.section not in self.sections():
                self.add_section(setting.section)
            if setting.option != None:
                self.set(setting.section, setting.option, setting.value)

    def read(self, filenames=None):
        """Read and parse a filename or a list of filenames.

        If filenames is None, it defaults to ._config_paths.  If a
        filename is not in ._config_paths, it gets appended to the
        list.  We also run os.path.expanduser() on the input filenames
        internally so you don't have to worry about it.

        Files that cannot be opened are silently ignored; this is
        designed so that you can specify a list of potential
        configuration file locations (e.g. current directory, user's
        home directory, systemwide directory), and all existing
        configuration files in the list will be read.  A single
        filename may also be given.

        Return list of successfully read files.
        """
        if filenames == None:
            filenames = [os.path.expanduser(p) for p in self._config_paths]
        else:
            if isinstance(filenames, basestring):
                filenames = [filenames]
            paths = [os.path.abspath(os.path.expanduser(p))
                     for p in self._config_paths]
            for filename in filenames:
                if os.path.abspath(os.path.expanduser(filename)) not in paths:
                    self._config_paths.append(filename)
        # Can't use super() because RawConfigParser is a classic class
        #return super(HookeConfigParser, self).read(filenames)
        return configparser.RawConfigParser.read(self, filenames)

    def _write_setting(self, fp, section=None, option=None, value=None,
                       **kwargs):
        """Print, if known, a nicely wrapped comment form of a
        setting's .help.
        """
        match = get_setting(self._default_settings, Setting(section, option))
        if match == None:
            match = Setting(section, option, value, **kwargs)
        match.write(fp, value=value, wrapper=self._comment_textwrap)

    def write(self, fp=None, comments=True):
        """Write an .ini-format representation of the configuration state.

        This expands on RawConfigParser.write() by optionally adding
        comments when they are known (i.e. for ._default_settings).
        However, comments are not read in during a read, so .read(x)
        .write(x) may `not preserve comments`_.

        .. _not preserve comments: http://bugs.python.org/issue1410680

        Examples
        --------

        >>> import sys, StringIO
        >>> c = HookeConfigParser()
        >>> instring = '''
        ... # Some comment
        ... [section]
        ... option = value
        ...
        ... '''
        >>> c._read(StringIO.StringIO(instring), 'example.cfg')
        >>> c.write(sys.stdout)
        [section]
        option = value
        <BLANKLINE>
        """
        local_fp = fp == None
        if local_fp:
            fp = open(os.path.expanduser(self._config_paths[-1]), 'w')
        if self._defaults:
            self._write_setting(fp, configparser.DEFAULTSECT,
                                help="Miscellaneous options")
            for (key, value) in self._defaults.items():
                self._write_setting(fp, configparser.DEFAULTSECT, key, value)
            fp.write("\n")
        for section in self._sections:
            self._write_setting(fp, section)
            for (key, value) in self._sections[section].items():
                if key != "__name__":
                    self._write_setting(fp, section, key, value)
            fp.write("\n")
        if local_fp:
            fp.close()

    def items(self, section, *args, **kwargs):
        """Return a list of tuples with (name, value) for each option
        in the section.
        """
        # Can't use super() because RawConfigParser is a classic class
        #return super(HookeConfigParser, self).items(section, *args, **kwargs)
        items = configparser.RawConfigParser.items(
            self, section, *args, **kwargs)
        for i,kv in enumerate(items):
            key,value = kv
            log = logging.getLogger('hooke') 
            try:
                setting = self._default_settings_dict[(section, key)]
            except KeyError, e:
                log.error('unknown setting %s/%s: %s' % (section, key, e))
                raise
            try:
                items[i] = (key, from_string(value=value, type=setting.type,
                                             count=setting.count))
            except ValueError, e:
                log.error("could not convert '%s' (%s) for %s/%s: %s"
                          % (value, type(value), section, key, e))
                raise
        return items

    def set(self, section, option, value):
        """Set an option."""
        setting = self._default_settings_dict[(section, option)]
        value = to_string(value=value, type=setting.type, count=setting.count)
        # Can't use super() because RawConfigParser is a classic class
        #return super(HookeConfigParser, self).set(section, option, value)
        configparser.RawConfigParser.set(self, section, option, value)

    def optionxform(self, option):
        """

        Overrides lowercasing behaviour of
        :meth:`ConfigParser.RawConfigParser.optionxform`.
        """
        return option


class TestHookeConfigParser (unittest.TestCase):
    def test_queue_safe(self):
        """Ensure :class:`HookeConfigParser` is Queue-safe.
        """
        from multiprocessing import Queue
        q = Queue()
        a = HookeConfigParser(default_settings=DEFAULT_SETTINGS)
        q.put(a)
        b = q.get(a)
        for attr in ['_dict', '_defaults', '_sections', '_config_paths',
                     '_default_settings']:
            a_value = getattr(a, attr)
            b_value = getattr(b, attr)
            self.failUnless(
                a_value == b_value,
                'HookeConfigParser.%s did not survive Queue: %s != %s'
                % (attr, a_value, b_value))
